source/client/ClientState.py

```diff
@@ -57,31 +65,81 @@ def reset(self):
+        #                           data: visible cards, which is processed in method:  in visible_scards[{...}]
         self.went_out = False
         self.discard_info = [None, 0]
-        
+
     def newCards(self, card_list):
         """Update the cards in hand"""
         for card in card_list:
             self.hand_cards.append(card)
 
-    def playCards(self, prepared_cards):
-        """Move cards from hand to visible"""
-        # First check that all the cards are in your hand
+    def playCards(self, prepared_cards, visible_scards=[{}], player_index=0):
+        """Move cards from hand to board if play follows rules, else inform what rule is broken."""
+
+        # First check that all the cards are in your hand.
+        self.player_index = player_index
         tempHand = [x for x in self.hand_cards]
         try:
             for card_group in prepared_cards.values():
                 for card in card_group:
                     tempHand.remove(card)
         except ValueError:
             raise Exception("Attempted to play cards that are not in your hand")
-        self.rules.canPlay(prepared_cards, self.played_cards, self.round)
-        for key, card_group in prepared_cards.items():
-            for card in card_group:
-                self.hand_cards.remove(card)
-                self.played_cards.setdefault(key, []).append(card)
-    
+        # Check ruleset to determine whether self.played_cards = all visible cards or cards that this client played.
+        if not self.rules.Shared_Board:
+            self.rules.canPlay(prepared_cards, self.played_cards, self.round)
+            for key, card_group in prepared_cards.items():
+                for card in card_group:
+                    self.hand_cards.remove(card)
+                    self.played_cards.setdefault(key, []).append(card)
+        elif self.rules.Shared_Board:
+            # Review Notes
+            # todo: move these to documentation
+            #  unlike in HandAndFoot, where self.played_cards was used to check rules.
+            # in Liverpool and other shared board games need to consider all of the played cards.
+            # Played cards (in deserialized form) are in visible_scards (in serialized form), which is obtained
+            # from controller.
+            # (Path taken by visible_scards:
+            #          Tableview gets the serialized cards every msec to keep display up to date,
+            #          In handview.update tableview.visible_scards list is passed to handview.visible_scards
+            #          No need to process this unless playing cards, in which case visible_scards passed
+            #          to controller and then to clientState, where only list item is deserialized and put in
+            #          dictionary self.played_cards
+            self.played_cards = self.rules.restoreRunAssignment(visible_scards[0], self.round)
```

Comment:
Also, as I have mentioned before, we can totally pass tempValues along in serialization and make this a straight deserialization step. I agree that having a rules class helper is good, largely because the dict structure is largely dealt with by the rules class (its like having the rules method for combining dicts)